### ISSUE TITLE
Feature request: Trigger a JS event on initialize/render

### ISSUE THREAD

**Issue by daattali** [ramnathv/htmlwidgets#250]
(Idea came from [this SO discussion](http://stackoverflow.com/questions/40920314/how-can-i-get-the-date-from-the-draggable-vertical-bar-in-timevis/41032504?noredirect=1#comment69303382_41032504))

Currently the only way (that I'm aware of?) to call some JS code on a widget once it initializes is to use the `htmlwidgets::onRender()` function. It's very useful, but sometimes it doesn't feel like the right thing to do for me because it feels a bit dirty adding a chunk of javascript code inside quotes inside an R function.

Also, suppose I create 5 different widgets, and I want the same JS code to run on each one after initialization. I'd have to add `%>% onRender(...)` on each of them.

I think it'd be useful to trigger a javascript event when an htmlwidget initializes, so that we could capture the event entirely in javascript

**Comment by ramnathv** [comment]
I think this is a good idea @daattali. I will wait for the others to weigh in.

**Comment by jcheng5** [comment]
Can you say more about what you envision happening in a global event handler like this? What would you use it for?

> Also, suppose I create 5 different widgets, and I want the same JS code to run on each one after initialization. I'd have to add %>% onRender(...) on each of them.

Or write an R function to abstract that away for you? I'd prefer that in at least some cases vs. inserting a standalone blob of JS that couples to every widget on the page, unless it's something very generic you're doing, but that's what I'm having a tough time imagining.

**Comment by daattali** [comment]
Yes, writing an R function to wrap around the onRender would be my current best approach rather than repeating it x times. But still, since I'm writing JS code that runs when a JS widget initializes, it'd feel more natural to do it entirely within JS.

To draw from the example that triggered this discussion, suppose someone wants to extend the `timevis` package for their usecase so that whenever a vertical timebar is dragged, it would send an input change event to Shiny. (You need to be familiar with `timevis` to understand this example...)  

@timelyportfolio 's answer was this:

```
timevis() %>%
  htmlwidgets::onRender(
"
function(el,x) {
  var tv = this.timeline;
  tv.on('timechanged', function(id, time) {
    Shiny.onInputChange(el.id + '_timechanged', {id:id, time:time})
  });
}
"
```

Which is great. And if I were to add this functionality to the package source, I would have also done it with similar code, and it would be implemented in the widget JS file. But as an end user wanting to supplement a 3rd party package like this,  I'd intuitively think I should be able to write this code in javascript following an event. 

It's not mandatory, as you said it can be done with onRender, but to me onRender is for one specific widget, whereas for a more generic case I'd want to do it in the language that widget is written in

**Comment by jcheng5** [comment]
Oh ok. So you'd subscribe to the event only for timevis widgets using
something like

$(document).on("htmlwidget:render", ".htmlwidget.timevis", ...)

Yeah, that makes total sense.

**Comment by timelyportfolio** [comment]
I am envisioning something much like https://shiny.rstudio.com/articles/js-events.html.

**Comment by daattali** [comment]
I didn't think about what would be the absolute best interaction, but yes
that was the general idea I wanted to support

---
R-Shiny Consultant
http://deanattali.com/shiny
